src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -974,6 +1042,65 @@ EXTERN_C NOINLINE void FASTCALL RhpGcPoll2(PInvokeTransitionFrame* pFrame)
     RhpWaitForGC2(pFrame);
 }
 
+#ifdef FEATURE_SUSPEND_REDIRECTION
+
+EXTERN_C NOINLINE void FASTCALL RhpSuspendRedirected()
+{
+    Thread* pThread = ThreadStore::GetCurrentThread();
+    CONTEXT* pCtx = pThread->GetRedirectionContext();
+
+    PInvokeTransitionFrame* frame =
+        (PInvokeTransitionFrame*)_alloca(PInvokeTransitionFrame_SaveAllRegs_SIZE);
+
+    frame->m_Flags = PROBE_SAVE_FLAGS_EVERYTHING;
+    frame->m_pThread = pThread;
+    frame->m_RIP = (void*)pCtx->GetIp();
+
+    UIntTarget* pPreservedRegs = frame->m_PreservedRegs;
+
+#ifdef TARGET_AMD64
+    frame->m_FramePointer = (void*)pCtx->Rbp;
+
+    pPreservedRegs[0] = pCtx->Rbx;
+    pPreservedRegs[1] = pCtx->Rsi;
+    pPreservedRegs[2] = pCtx->Rdi;
+    pPreservedRegs[3] = pCtx->R12;
+    pPreservedRegs[4] = pCtx->R13;
+    pPreservedRegs[5] = pCtx->R14;
+    pPreservedRegs[6] = pCtx->R15;
+
+    pPreservedRegs[7] = pCtx->Rsp;
+
+    pPreservedRegs[8] = pCtx->Rax;
+    pPreservedRegs[9] = pCtx->Rcx;
+    pPreservedRegs[10] = pCtx->Rdx;
+    pPreservedRegs[11] = pCtx->R8;
+    pPreservedRegs[12] = pCtx->R9;
+    pPreservedRegs[13] = pCtx->R10;
+    pPreservedRegs[14] = pCtx->R11;
+#elif defined(TARGET_ARM64)
+    frame->m_FramePointer = (void*)pCtx->Fp;
+    // X19-X28 
+    memcpy(&pPreservedRegs[0], &pCtx->X19, sizeof(uint64_t) * 10);
+    // SP
+    pPreservedRegs[10] = pCtx->Sp;
+    // X0-X18 
+    memcpy(&pPreservedRegs[11], &pCtx->X0, sizeof(uint64_t) * 19);
+    // LR
+    pPreservedRegs[30] = pCtx->Lr;
+#elif
+    ASSERT_UNCONDITIONALLY("NYI for this arch");
+#endif
+
+    pThread->WaitForGC(frame);
+
+    // restore execution at interrupted location
```

Comment:
Do we need to propagate the register values back in case the GC updated them?

Reply:
Right. Forgot about that part. :)

Reply:
I am also thinking about making it possible to seed stackwalks with CONTEXT* on Windows.
That may allow avoiding copying the registers here and in Pal, and remove alloca. 
Probably not in this PR though, since I want to unblock libraries testing.

Reply:
Actually not copying the registers solves both problems - no redundant copying and no need to reflect back the changes.